agg-ast/schema_derivation/src/schema_derivation.rs

```diff
@@ -429,11 +429,157 @@ impl DeriveSchema for Stage {
+                    match project_item {
+                        ProjectItem::Inclusion => {
+                            schemas.insert(schema.clone());
+                        }
+                        ProjectItem::Exclusion => return Schema::Missing,
+                        ProjectItem::Assignment(e) => {
+                            let schema = e.derive_schema(state);
+                            if let Ok(s) = schema {
+                                schemas.insert(s);
+                            }
+                        }
+                    }
+                }
+                match schema {
+                    // if the current node is an Array, any child schemas will have to come from nested documents in this array.
+                    // evaluate the inner type of the array for any nested fields, then wrap the resulting schema in an Array
+                    Schema::Array(a) => {
+                        let inner_schema =
+                            get_schema_for_project_node(node, Some(a.as_ref()), state);
+                        schemas.insert(Schema::Array(Box::new(inner_schema)));
+                    }
+                    // similarly, anyofs can be part of a nested path if any of schemas are documents (or arrays of documents).
+                    // we evaulate each subtype to verify if it contains any of the children nodes
+                    Schema::AnyOf(ao) => {
+                        ao.iter().for_each(|ao_schema| {
+                            let schema = get_schema_for_project_node(node, Some(ao_schema), state);
+                            if schema != Schema::Unsat {
+                                schemas.insert(schema);
+                            }
+                        });
+                    }
+                    // documents are where we actually evaluate the children schemas to see if their paths are present in the document.
+                    // we iterate through each child, and build a document from the resulting key-schema pairs.
+                    d @ Schema::Document(_) => {
+                        let d = promote_missing(d);
+                        let mut path_document = Schema::Document(Document::empty());
+                        for (child_field, child_node) in node.children_nodes.iter() {
+                            let child_schema = get_schema_for_project_node(
+                                child_node,
+                                d.get_key(child_field),
+                                state,
+                            );
+                            insert_required_key_into_document(
+                                &mut path_document,
+                                child_schema,
+                                vec![child_field.clone()],
+                                true,
+                            );
+                        }
+                        schemas.insert(path_document);
+                    }
+                    _ => {}
```

Comment:
Sorry I'm a bit out of date, but do we expand `Any` to an `AnyOf` before this? Is that relevant?

Reply:
This is a good callout, this is correct, we need to do this. This also is making me realize we should add tests to handle additionalproperties correctly. Will add